### PR TITLE
Prod 786 generator add specs fields

### PR DESCRIPTION
**Pull request by renanbronchart**

### CODE REVIEW

File: tests/unit/GoFields/GoFieldInput.spec.js [hunk 1]
```diff
@@ -1,25 +1,47 @@
...
+
   // Install
   it('should be installed', () => {
     expect(localVue.options.components['GoFieldInput']).toBeTruthy()
   })
 
-  // Render testing
-  // renderCases.forEach(renderCase => {
-  //   const [fromComp, toHtml] = renderCase
+  it('should have errors bag present', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldInput.spec.js]: Replace by : `should have vee-validate errors register`

File: tests/unit/GoFields/GoFieldInput.spec.js [hunk 2]
```diff
@@ -1,25 +1,47 @@
...
+    expect(wrapper.vm.errors.items).toBeDefined()
+  })
+
+  it('should have not vee-validate errors register', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldInput.spec.js]: should have not registered vee-validate error(s)

File: tests/unit/GoFields/GoFieldInput.spec.js [hunk 3]
```diff
@@ -1,25 +1,47 @@
...
+    expect(wrapper.vm.errors.items).toEqual([])
+  })
+
+  it('should have input value equal to v-model pass in props', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldInput.spec.js]: `should have input value equal to v-model passed in (the name of the prop) prop`

File: tests/unit/GoFields/GoFieldInput.spec.js [hunk 4]
```diff
@@ -1,25 +1,47 @@
...
+    expect(wrapper.vm.valueInput).toEqual(wrapper.vm.value)
+  })
 
-  //   it(`"${fromComp}" should render "${toHtml}"`, () => {
-  //     const wrapper = mount({ template: fromComp }, { localVue })
-  //     expect(wrapper.html()).toBe(toHtml)
-  //   })
-  // })
+  it('should emit input event with input value', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldInput.spec.js]: should emit input event with the input value

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 5]
```diff
@@ -1,25 +1,87 @@
...
-  // Render testing
-  // renderCases.forEach(renderCase => {
-  //   const [fromComp, toHtml] = renderCase
+  it('should be default options component', () => {
+    expect(buttonSubmit.is('button')).toBe(true)
+
+    expect(buttonSubmit.attributes('type')).toBe('submit')
+  })
+
+  it('should have empty attributes which no fill', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: should have empties attributes which no fill

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 6]
```diff
@@ -1,25 +1,87 @@
...
+    expect(buttonSubmit.attributes('value')).toBe(undefined)
+    expect(buttonSubmit.attributes('name')).toBe(undefined)
+    expect(buttonSubmit.attributes('disabled')).toBe(undefined)
+    expect(buttonSubmit.attributes('class')).toBe(undefined)
+  })
+
+  it('should emit submit event when click on submit button', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: should emit submit event when there is a click event on submit button

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 7]
```diff
@@ -1,25 +1,87 @@
...
-  //     const wrapper = mount({ template: fromComp }, { localVue })
-  //     expect(wrapper.html()).toBe(toHtml)
-  //   })
-  // })
+    expect(wrapper.emitted().submit).toBeTruthy()
+    expect(wrapper.emitted().submit.length).toBe(1)
+    expect(wrapper.emitted().submit[0][0].type).toBe('click')
+  })
+
+  it('should emit submit event when enter on submit button', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: should emit submit event when there is a enter key event on submit button

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 8]
```diff
@@ -1,25 +1,87 @@
...
+    buttonSubmit.trigger('keyup.enter')
+
+    expect(wrapper.emitted().submit).toBeTruthy()
+    expect(wrapper.emitted().submit.length).toBe(1)
+    expect(wrapper.emitted().submit[0][0].type).toBe('keyup')
+    expect(wrapper.emitted().submit[0][0].keyCode).toBe(13)
+  })
+
+  describe('when schema is fill', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: when schema is filled

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 9]
```diff
@@ -1,25 +1,87 @@
...
+      })
+
+      buttonSubmit = wrapper.find({ref: 'submit'})
+    })
+
+    it('should be tag component specified in schema props', () => {
+      expect(buttonSubmit.is('input')).toBe(true)
+    })
+
+    it('should have attributes which specified in props', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: should have attributes which are specified in props

File: tests/unit/GoFields/GoFieldSubmit.spec.js [hunk 10]
```diff
@@ -1,25 +1,87 @@
...
+          schema,
+          isDisabled: true
+        },
+        localVue,
+      })
+
+      buttonSubmit = wrapper.find({ref: 'submit'})
+    })
+
+    it('should be tag component specified in schema props', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFields/GoFieldSubmit.spec.js]: I don't understand.

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+    })
+  })
+
+  //////////
+  // Computed properties
+  //////////
+
+  // computed - schemaFields
+  describe('computed - schemaFields', () => {
+    it('should return empty array if not any field is filled', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an empty array if the fields are not filled

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+      expect(wrapper.vm.schemaFields).toEqual([])
+    })
+
+    describe('when fields are fill in schema', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when fields are filled in schema

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+                type: 'input',
+                inputType: 'text',
+                inputName: 'first_name',
+                model: 'first_name',
+            }],
+          }
+        })
+      })
+
+      it('should return array of fields according to schemaFields', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an array of fields according to schemaFields

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+                    inputName: 'last_name',
+                    model: 'last_name',
+                }],
+              }
+            ]
+          }
+        })
+      })
+
+      it('should return empty array', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an empty array

File: tests/unit/GoFields/GoFieldSubmit.spec.js (hunk 5 above)
Comment (benjaminrobert) [reply tests/unit/GoFields/GoFieldSubmit.spec.js]: should have empty attributes which are not filled

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+        expect(wrapper.vm.schemaFields).toEqual([])
+      })
+    })
+  })
+
+
+  // computed - groups
+  describe('computed - groups', () => {
+    it('should return empty array if not any group is filled', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an empty array if the is no filled group

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+      expect(wrapper.vm.schemaGroups).toEqual([])
+    })
+
+    describe('when fields are fill in schema without sub group', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when fields are filled in schema without sub group

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+                type: 'input',
+                inputType: 'text',
+                inputName: 'first_name',
+                model: 'first_name',
+            }],
+          }
+        })
+      })
+
+      it('should return empty array', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an empty array

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+        expect(wrapper.vm.schemaGroups).toEqual([])
+      })
+    })
+
+    describe('when groups are fill in schema', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when groups are filled in schema

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+                    inputName: 'last_name',
+                    model: 'last_name',
+                }],
+              }
+            ]
+          }
+        })
+      })
+
+      it('should return array of groups according to groupsFields', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should return an array of groups according to groupsFields

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+          }
+        })
+      })
+
+      it('should return array of fields according to schemaFields', () => {
+        expect(wrapper.vm.schemaFields).toEqual(wrapper.vm.schema.fields)
+      })
+    })
+
+    describe('when groups are fill in schema', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the groups are filled in schema

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+    })
+
+    describe('when hidden property of field is set to true', () => {
+      it('should field is not visible', () => {
+        field.hidden = true
+        expect(wrapper.vm.fieldVisible(field)).toBe(false)
+      })
+    })
+
+    describe('when hidden property of field is set to false', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the hidden property of field is setted to false

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+
+  // methods - hasErrorOnFields
+  describe('methods - fieldVisible', () => {
+    let refs = {}
+
+    it('should return no errors by default', () => {
+      expect(wrapper.vm.hasErrorOnFields(refs)).toBe(false)
+    })
+
+    describe('when ref field are present and have not errors', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the ref fields are present and have no errors

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+      it('should have not error on field', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should have no error on the field

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 24]
```diff
@@ -0,0 +1,349 @@
...
+              hasErrors: false
+            }
+          ]
+        }
+
+        expect(wrapper.vm.hasErrorOnFields(refs)).toBe(false)
+      })
+    })
+
+    describe('when ref field are present and some have errors', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the ref fields are present and have some errors

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 25]
```diff
@@ -0,0 +1,349 @@
...
+      it('should have errors on field', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should have errors on the field

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 26]
```diff
@@ -0,0 +1,349 @@
...
+              hasErrors: false
+            }
+          ]
+        }
+
+        expect(wrapper.vm.hasErrorOnFields(refs)).toBe(true)
+      })
+    })
+
+    describe('when ref group are present and have not error', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the ref groups are present and have no error

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 27]
```diff
@@ -0,0 +1,349 @@
...
+      it('should have not error on field', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should have no error on the field

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 28]
```diff
@@ -0,0 +1,349 @@
...
+              hasErrors: false
+            }
+          ]
+        }
+
+        expect(wrapper.vm.hasErrorOnFields(refs)).toBe(false)
+      })
+    })
+
+    describe('when ref group are present and have errors', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the ref groups are present and have errors

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 29]
```diff
@@ -0,0 +1,349 @@
...
+      it('should have errors on field', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should have errors on the field

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 30]
```diff
@@ -0,0 +1,349 @@
...
+              hasErrors: false
+            }
+          ]
+        }
+
+        expect(wrapper.vm.hasErrorOnFields(refs)).toBe(true)
+      })
+    })
+
+    describe('when ref field and ref group are present and have errors', () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: when the ref field and the ref group are present and have errors

File: tests/unit/GoFormGenerator/GoFormGenerator.spec.js [hunk 31]
```diff
@@ -0,0 +1,349 @@
...
+  })
+
+
+  //////////
+  // Watch properties
+  //////////
+
+  // watch - value
+  describe('Watch - value', () => {
+    it('should emit input value when change value', async () => {
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGenerator.spec.js]: should emit the input value when the value is changed

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 32]
```diff
@@ -0,0 +1,152 @@
...
+    placeholder: 'Your last name',
+    required: true,
+    validations: 'required'
+  },
+}
+
+// render cases for computed hasErrors
+export const renderCaseHasErrors = [
+  [
+    'when there are no vee-validate and data base errors',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are no vee-validate and database errors

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+      propsData: basePropsData,
+      computed: {
+        hasErrorVeeValidate: () => true,
+        fieldModelErrors: () => [],
+      }
+    },
+    true
+  ],
+  [
+    'when there are database errors but display fields error is not activated',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are database errors but the display of the fields error is not activated

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 34]
```diff
@@ -0,0 +1,152 @@
...
+      propsData: basePropsData,
+      computed: {
+        hasErrorVeeValidate: () => false,
+        fieldModelErrors: () => ['required'],
+      }
+    },
+    false
+  ],
+  [
+    'when there are database errors and display fields error is activated',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are database errors and the display of the fields error is activated

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 35]
```diff
@@ -0,0 +1,152 @@
...
+      propsData: basePropsData,
+      computed: {
+        hasErrorVeeValidate: () => false,
+        fieldModelErrors: () => ['required'],
+      }
+    },
+    true
+  ],
+  [
+    'when there are database errors and vee-validate errors. Display fields error is activated',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are database errors and vee-validate errors and the display of the fields error is activated

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 36]
```diff
@@ -0,0 +1,152 @@
...
+    },
+    true
+  ]
+]
+
+// render cases for computed hasErrors
+export const renderCaseFieldModelErrors = [
+  [
+    // describe
+    'when there are errors which are not according to field model',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are errors which are not according to the field model

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 37]
```diff
@@ -0,0 +1,152 @@
...
+        { 'code': 'required', 'source': '/data/attributes/first_name' },
+        { 'code': 'email', 'source': '/data/attributes/first_name' },
+      ],
+    },
+    // expect value
+    []
+  ],
+  [
+    // describe
+    'when there is one error which is according to field model',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there is one error which is according to the field model

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js [hunk 38]
```diff
@@ -0,0 +1,152 @@
...
+        { 'code': 'required', 'source': '/data/attributes/last_name' },
+        { 'code': 'email', 'source': '/data/attributes/first_name' },
+      ],
+    },
+    // expect value
+    ['required']
+  ],
+  [
+    // describe
+    'when there is are all errors which are according to field model',
```
Comment (MatthiasHe) [inline tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: Not sure to understand this

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js (hunk 36 above)
Comment (benjaminrobert) [reply tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there are errors which are not related to the field model

File: tests/unit/GoFormGenerator/GoFormGroup.render-cases.js (hunk 37 above)
Comment (benjaminrobert) [reply tests/unit/GoFormGenerator/GoFormGroup.render-cases.js]: when there is one error which is related to the field model

**Pull request merged**